File: src/protobuf/examples/list_people.py
#! /usr/bin/python

# see readme.txt for information and build instructions.

import addressbook_pb2
import sys

# iterates though all people in the addressbook and prints info about them.
def listpeople(address_book):
  for person in address_book.person:
    print "person id:", person.id
    print "  name:", person.name
    if person.hasfield('email'):
      print "  e-mail address:", person.email

    for phone_number in person.phone:
      if phone_number.type == addressbook_pb2.person.mobile:
        print "  mobile phone #:",
      elif phone_number.type == addressbook_pb2.person.home:
        print "  home phone #:",
      elif phone_number.type == addressbook_pb2.person.work:
        print "  work phone #:",
      print phone_number.number

# main procedure:  reads the entire address book from a file and prints all
#   the information inside.
if len(sys.argv) != 2:
  print "usage:", sys.argv[0], "address_book_file"
  sys.exit(-1)

address_book = addressbook_pb2.addressbook()

# read the existing address book.
f = open(sys.argv[1], "rb")
address_book.parsefromstring(f.read())
f.close()

listpeople(address_book)

File: src/protobuf/examples/list_people.cc
// see readme.txt for information and build instructions.

#include <iostream>
#include <fstream>
#include <string>
#include "addressbook.pb.h"
using namespace std;

// iterates though all people in the addressbook and prints info about them.
void listpeople(const tutorial::addressbook& address_book) {
  for (int i = 0; i < address_book.person_size(); i++) {
    const tutorial::person& person = address_book.person(i);

    cout << "person id: " << person.id() << endl;
    cout << "  name: " << person.name() << endl;
    if (person.has_email()) {
      cout << "  e-mail address: " << person.email() << endl;
    }

    for (int j = 0; j < person.phone_size(); j++) {
      const tutorial::person::phonenumber& phone_number = person.phone(j);

      switch (phone_number.type()) {
        case tutorial::person::mobile:
          cout << "  mobile phone #: ";
          break;
        case tutorial::person::home:
          cout << "  home phone #: ";
          break;
        case tutorial::person::work:
          cout << "  work phone #: ";
          break;
      }
      cout << phone_number.number() << endl;
    }
  }
}

// main function:  reads the entire address book from a file and prints all
//   the information inside.
int main(int argc, char* argv[]) {
  // verify that the version of the library that we linked against is
  // compatible with the version of the headers we compiled against.
  google_protobuf_verify_version;

  if (argc != 2) {
    cerr << "usage:  " << argv[0] << " address_book_file" << endl;
    return -1;
  }

  tutorial::addressbook address_book;

  {
    // read the existing address book.
    fstream input(argv[1], ios::in | ios::binary);
    if (!address_book.parsefromistream(&input)) {
      cerr << "failed to parse address book." << endl;
      return -1;
    }
  }

  listpeople(address_book);

  // optional:  delete all global objects allocated by libprotobuf.
  google::protobuf::shutdownprotobuflibrary();

  return 0;
}
